Build: add support for native TypeScript

jsnajdr opened the pull request:
Adds support for compiling native `.ts` and `.tsx` syntax to build scripts. Spinoff from #28465 that uses this tooling to implement a `react-i18n` package written as native TypeScript source.

First of all, we're adding the `@babel/preset-typescript` preset to the `babel-preset-default` so that the Gutenberg Babel config can parse the TypeScript syntax. With this support, Babel will treat the type annotations as comments and will ignore them, processing the file as standard JavaScript.

The TypeScript preset adds one new constraint to how we use Babel: we need to pass the `filename` now, in addition to the source string. Both when calling the Babel parser and when calling the Babel transformer. The TypeScript preset uses the `filename` to determine whether it should activate the TypeScript mode: it does so only for `.tsx?` files.

When called by webpack or similar tools, Babel gets the `filename` argument. In the Gutenberg codebase, we need to add it to unit tests and also to `docgen` which uses the Babel parser.

Next, we add support for `.tsx?` files to the build scripts that search for source files to transpile and do the actual transpilation.

Last, we need to update the `update-api-docs.js` script so that when it encounters an API docs token in a `README.md` file:
```
<!-- START TOKEN(Autogenerated API docs) -->
```
it finds the `src/index.ts` as the package's default source file. Until now, it was looking only for `src/index.js`.

jsnajdr committed:
babel-preset-default: add support for TypeScript parsing

jsnajdr committed:
docgen: pass filename to Babel parser

jsnajdr committed:
babel-plugin-makepot: pass filename to Babel in unit tests

jsnajdr committed:
Build: transpile TS files

jsnajdr committed:
Update API Docs: default package source file can have TS extension

nosolosw commented on packages/docgen/src/engine.js:
```diff
@@ -19,25 +18,21 @@ const getExportTokens = ( ast ) =>
-			)
-		)
+	if ( ! path || ! code ) {
+		return null;
```
This changes the output of `engine` for consumers. We should either maintain the output or adapt https://github.com/WordPress/gutenberg/blob/master/packages/docgen/src/index.js#L54 to work with nulls.

jsnajdr replied:
OK, I reverted the return value of `processFile` to always be an object. In case of error, it will return `{ ast: null, tokens: null, ir: [] }`. That's what the consumers and also what the tests expect.

nosolosw commented on packages/docgen/src/engine.js:
```diff
@@ -2,15 +2,14 @@
 
-const getAST = ( source ) => {
-	return babel.parse( source || '' ).program;
+const getAST = ( source, filename ) => {
```
Reading the issue description I gather we use this to signal the `babel-preset-typescript` babel is processing a tsx file, but I haven't found info about that [here](https://babeljs.io/docs/en/babel-parser) or [here](https://babeljs.io/docs/en/babel-preset-typescript). Just curiosity, but it'd be nice to have some pointers to this.

jsnajdr replied:
Here we're not using the `@babel/parser` module directly, but indirectly through the `@babel/core` module which exports a `parse` function. `@babel/core` loads the Babel config will all the plugins and presets, and then exposes a configured parser.

This `parse` method is documented [here](https://babeljs.io/docs/en/babel-core#parsesync) and the `filename` option is documented [here](https://babeljs.io/docs/en/options#filename).

If the `filename` option is not present, the parsing will fail with error:
> Error: [BABEL] unknown: Preset /* your preset */ requires a filename to be set when babel is called directly,
> ```
> babel.transform(code, { filename: 'file.ts', presets: [/* your preset */] });
> ```
> See https://babeljs.io/docs/en/options#filename for more information.
>    at validateIfOptionNeedsFilename (/Users/jsnajdr/src/gutenberg/node_modules/@babel/core/lib/config/full.js:286:11)

nosolosw commented:
Changes to `docgen` look fine, although I left a comment I'd like us to address.

I have a larger question about this PR: does it make sense to process TSX files with `docgen`? This package is used to generate API docs automatically based on JSDoc syntax. If exported entities are not going to have data about their input & output (names, type, description) in a way docgen can process it, is still useful that we generate API docs? It sounds like we shouldn't use `docgen` for TSX files or, alternatively, add support for typescript syntax within docgen itself.

cc @saramarcondes as per your work on https://github.com/WordPress/gutenberg/pull/28615 related to this as well.

saramarcondes commented:
Agreeing with you, @nosolosw, that that should be the general goal; however, this might be a fine stop-gap for now, to at least have parameters documented without their types.

I still need to sit down and figure out how TSDoc works and how we might be able to deploy its usage in the Gutenberg repository. I'll do that today and get back to y'all on whether it's something simple we could include in this PR if it's complicated enough to need a new PR.

nosolosw commented:
> this might be a fine stop-gap for now, to at least have parameters documented without their types.

I may be missing some context, as I've only looked at [the first file we're introducing](https://github.com/WordPress/gutenberg/pull/28465/files#diff-d9fe05747d9a2705393b29af2553c0e4ed4ee536495953647f8b1e06b6404db3R111). It doesn't look like the exported entities in the typescript files are going to have any more info than their description. So it's not that the API docs won't have the types, but that they won't have anything about input and output data (param names, types, or description and return type and description) ― there are no JSDoc `@params` or `@return` tags to extract that data from.

My point is that it should be fine that certain typescript packages or files don't use `docgen` until we find a solution that works for them (either a replacement or an update).

saramarcondes commented:
>  there are no JSDoc @params or @return tags to extract that data from.

Ah, I hadn't looked at the TypeScript files but you're right, that will be an issue.

There's nothing preventing us from documenting those arguments, however. We can't add the types but we can add `@param` and `@return` tags to those functions, at least then we'd get descriptions.

But if you think we should just hand-write those documentations then I say go for it, if the types are that important to the docs.

An update on TSDoc, the main project that seems to exist for extracting TSDoc comments into Markdown is `API Extractor`, however it does not support "lazy imports" in the generated `.d.ts` files, which `react-i18n` does use (`withI18n`'s return type is inferred as a React.FunctionComponent which gets transformed by the TypeScript compiler to `import('react').FunctionComponent`; it's easy to see this if you run `build:packages` and inspect the output `.d.ts` files. We also use this all over the project to share types in JSDoc comments). [There's a work around](https://github.com/microsoft/rushstack/issues/2140#issuecomment-687524094) but it's not viable in my opinion for the reasons shared in this comment: https://github.com/microsoft/rushstack/issues/2140#issuecomment-687710391

In any case, I think what you said is correct @nosolosw: we either hand write the documentation or sacrifice the types (in which case we'd need to update the code comments to include parameter and return descriptions).

jsnajdr commented:
The only thing that JSDoc comments in TypeScript sources don't have are the types, because they are defined in the source and specifying them in JSDoc would be duplication.

I can imagine the `docgen` tool extracting the type annotations from the TypeScript AST and writing it in the output.

Processing this:
```ts
/**
 * Sum two numbers.
 *
 * @param a The first number
 * @param b The second number
 * @returns Their sum
 */
function sum( a: number, b: number ): number {
  return a + b;
}
```

Is only slightly more difficult than processing this classic JSDoc:
```js
/**
 * Sum two numbers.
 *
 * @param {number} a The first number
 * @param {number} b The second number
 * @returns {number} Their sum
 */
function sum( a, b ) {
  return a + b;
}
```

The question is what to do when the types are not as neat as above. For example, we didn't have to specify the return type of `sum` -- it's easily inferred. But `docgen` wouldn't know it.

saramarcondes commented:
@jsnajdr I spent a good chunk of time one weekend trying to re-write docgen using the TypeScript parser, which would make that possible, but it's a big endeavor, it's not a simple change.

I doubt the babel parser preserves the types in the AST, so how would we extract it without converting to the TypeScript parser?

jsnajdr commented:
> I doubt the babel parser preserves the types in the AST, so how would we extract it without converting to the TypeScript parser?

The Babel parser can parse both TypeScript and Flow type annotations and add them to the AST. Example from astexplorer.net:
<img width="1196" alt="Screenshot 2021-02-09 at 15 27 09" src="https://user-images.githubusercontent.com/664258/107377598-705e0200-6aeb-11eb-8c45-2376471adbe3.png">

saramarcondes commented:
> The Babel parser can parse both TypeScript and Flow type annotations and add them to the AST

That's incredible good news! Is that something we should implement as part of this PR? I'm happy to help with that if you'd like.

jsnajdr commented:
> Is that something we should implement as part of this PR?

It's definitely something that we should implement, but I believe it's not a good topic for this PR 🙂 This PR is already a spinoff from something larger (#28465) and is supposed to be reviewed and merged quickly.

Let's work on this either in the #28465 branch, or in a completely new one.

gziolo commented:
If the only limitation for now is that you need to duplicate TS types in JSDoc comments then we should land it as is. We don't use TS at scale so it falls in the same bucket as using ESLint disable for complex TS types in JSDoc comments. It's inconvenient but it requires more work to sort it out. Let's keep those PR small and open an issue that tracks next steps.

jsnajdr commented:
> I may be missing some context, as I've only looked at the first file we're introducing. It doesn't look like the exported entities in the typescript files are going to have any more info than their description.

@nosolosw I added some `@param` and `@return` tags to the `react-i18n` docs in the #28465 branch. No types, just param names and descriptions. This is how the resulting Markdown looks like:

<img width="858" alt="Screenshot 2021-02-10 at 10 20 17" src="https://user-images.githubusercontent.com/664258/107490349-12342c00-6b8a-11eb-95b2-95a5b60ff504.png">

The `(unknown type)` annotations are ugly (do we really need to add them at all?) but otherwise the docs are acceptable at least.

I believe I addressed all your review comments. Can you have another look?

gziolo commented:
> The (unknown type) annotations are ugly (do we really need to add them at all?) but otherwise the docs are acceptable at least.

It's a bug that @sarayourfriend has addressed in another PR that should land soon.

nosolosw commented:
The changes related to generating API docs are fine :+1: I'm not deeply familiar with the build process, so rather let other people comment on those.

gziolo reviewed (approved):
Let's give it a try and iterate 😃
